// @ctrl/ngx-csv/esm2015/csv.directive.js
import { Directive, HostBinding, HostListener, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { blob, buildURI } from './util';
export class CsvDirective {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
        /** the body of the csv */
        this.data = [];
        /** Set the seperator between values */
        this.delimiter = ',';
        /** adds a Byte order mark to setup the csv as UTF-8  */
        this.uFEFF = true;
        /** filename */
        this.download = 'data.csv';
        this.target = this.isIEBrowser() ? '' : '_blank';
    }
    /** Set the filename of the csv. Default is `data.csv` */
    set filename(a) {
        this.download = a;
    }
    onClick() {
        // IE handling
        if (this.isIEBrowser()) {
            const file = blob(this.data, this.uFEFF, this.headers, this.delimiter);
            window.navigator.msSaveBlob(file, this.download);
        }
    }
    isIEBrowser() {
        return !!window.navigator.msSaveOrOpenBlob;
    }
    ngOnChanges() {
        this.href = this.sanitizer.bypassSecurityTrustResourceUrl(buildURI(this.data, this.uFEFF, this.headers, this.delimiter));
    }
}
CsvDirective.decorators = [
    { type: Directive, args: [{ selector: '[csvLink]' },] }
];
CsvDirective.ctorParameters = () => [
    { type: DomSanitizer }
];
CsvDirective.propDecorators = {
    data: [{ type: Input }],
    headers: [{ type: Input }],
    delimiter: [{ type: Input }],
    filename: [{ type: Input }],
    uFEFF: [{ type: Input }],
    href: [{ type: HostBinding }],
    download: [{ type: HostBinding }],
    target: [{ type: Input }, { type: HostBinding }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
